quinn-proto/src/endpoint.rs

```diff
@@ -690,35 +690,27 @@ impl Endpoint {
 
     /// Handle a timer expiring
     pub fn timeout(&mut self, now: u64, conn: ConnectionHandle, timer: Timer) {
+        let mut mux = EndpointMux::new(&mut self.ctx, conn, self.connections[conn.0].side());
+        let closed = self.connections[conn.0].timeout(&mut mux, now, timer);
         match timer {
             Timer::Close => {
                 self.ctx.io.push_back(Io::TimerStop {
                     connection: conn,
                     timer: Timer::Idle,
                 });
                 self.ctx.events.push_back((conn, Event::ConnectionDrained));
-                if self.connections[conn.0].app_closed {
+                if closed {
                     self.forget(conn);
-                } else {
-                    self.connections[conn.0].state = State::Drained;
                 }
             }
             Timer::Idle => {
-                let mut mux =
-                    EndpointMux::new(&mut self.ctx, conn, self.connections[conn.0].side());
-                self.connections[conn.0].close_common(&mut mux, now);
-                self.connections[conn.0].state = State::Draining;
                 self.ctx
                     .events
                     .push_back((conn, ConnectionError::TimedOut.into()));
                 self.dirty_conns.insert(conn); // Ensure the loss detection timer cancellation
                                                // goes through
```

Comment:
Probably best to do this unconditionally on any timeout; loss detection and close timeout can both incur I/O events, after all.

Reply:
On second thought, a close timeout should *not* cause a new entry in `dirty_conns`, because the only I/O it might incur is queued proactively rather than needing to be polled out, and the connection might cease to exist afterwards.

Reply:
And, in fact, the comment on this specific line is obsolete; we *only* need to mark this connection dirty for the loss detection timer.

Reply:
As of tonight's refactoring, we once again need to mark dirty for potential timer updates.